include/pybind11/stl.h

```diff
@@ -305,6 +314,58 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static handle cast(const std::filesystem::path& path, return_value_policy, handle) {
+        PyObject* py_str = [](const auto& w) {
+            using char_type = std::filesystem::path::value_type;
+            if constexpr (std::is_same_v<char_type, char>) {
+                return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+            } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+                return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+            }
+        } (path.native());
+        if (!py_str) {
+            return nullptr;
+        }
+        return module::import("pathlib").attr("Path")(reinterpret_steal<object>(py_str))
```

Comment:
I don't think there's much to be gained from that, and this feels just annoying for end users who may want to actually perform filesystem ops using the Python object (then they'd need to cast the thing to a Path)?  Also std::filesystem::path does come with filesystem access (non-member) functions, so that makes it closer to Path than to PurePath anyways.
I think PurePath is really only relevant for the case of creating Windows paths on Linux and vice-versa, but std::filesystem::path doesn't allow that AFAICT...

Reply:
Good point.